File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.c
int lengthOfLongestSubstring(char * s){
    int lastIndex[128]; // ASCII size
    int i,end;
    for (i = 0; i < 128; i++) {
        lastIndex[i] = -1;
    }

    int maxLen = 0;
    int start = 0;

    for (end = 0; s[end] != '\0'; end++) {
        char currentChar = s[end];

        // If character has been seen and is inside current window
        if (lastIndex[currentChar] >= start) {
            start = lastIndex[currentChar] + 1; // move start
        }

        // Update the last index of the character
        lastIndex[currentChar] = end;

        int windowLen = end - start + 1;
        if (windowLen > maxLen) {
            maxLen = windowLen;
        }
    }

    return maxLen;
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        int maxLen = 0;

        HashMap<Character, Integer> window = new HashMap<>();

        int l = 0;

        for(int r = 0; r < s.length(); r++){


            window.put(s.charAt(r), window.getOrDefault(s.charAt(r), 0) + 1);

            while(window.get(s.charAt(r)) > 1){

                char remove = s.charAt(l);
                
                window.put(remove, window.get(remove) - 1);

                if(window.get(remove) == 0) window.remove(remove);

                l++;
                
            }

            maxLen = Math.max(maxLen, window.size());
        }

        return maxLen;
    }
}